dev/PushNotifications/PushNotificationUtility.cpp

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include "pch.h"
+#include "externs.h"
+#include "PushNotificationUtility.h"
+
+const std::wstring ConvertByteArrayToWideString(unsigned int payloadLength, _In_reads_(payloadLength) byte* payload)
+{
+    int size = MultiByteToWideChar(
+        CP_UTF8,
+        0,
+        reinterpret_cast<PCSTR>(payload),
+        payloadLength,
+        nullptr,
+        0);
+    THROW_LAST_ERROR_IF(size == 0);
+
+    std::wstring payloadAsWideString(size, 0);
+    size = MultiByteToWideChar(
+        CP_UTF8,
+        0,
+        reinterpret_cast<PCSTR>(payload),
+        payloadLength,
+        &payloadAsWideString[0],
+        size);
+    THROW_LAST_ERROR_IF(size == 0);
+
+    return payloadAsWideString;
+}
+
+void ProtocolLaunchHelper(unsigned int payloadLength, _In_reads_(payloadLength) byte* payload)
+{
+    // Command line format: ----WindowsAppRuntimePushServer:-Payload:"<payloadAsEscapedUriFormat>"
+    std::wstring commandLine = L"----WindowsAppRuntimePushServer:-Payload:\"";
+
+    // Escape special characters to follow command line standards for any app activation type in AppLifecycle
+    // (See AppInstance.cpp and Serialize() from other activation types)
+    std::wstring payloadAsWideString = ConvertByteArrayToWideString(payloadLength, payload);
```

Comment:
I can't remember if we talked about this before or not,  but I believe we would be better served not creating some new payload model.  Instead we should create a contractId for the push server, and use the already existing infrastructure to pass properties through the query string.  This might be an example of a fragile design since it requires continued upkeep to make it special.

Reply:
@jonwis /@aeloros - I see howard introducing a lot of string conversion APIs in dev/common: https://github.com/microsoft/WindowsAppSDK/pull/1513. Once the PR merges some of these conversion APIs in my .h/.cpp will be removed. 

@aeloros - "Instead we should create a contractId for the push server, and use the already existing infrastructure to pass properties through the query string. This might be an example of a fragile design since it requires continued upkeep to make it special" -  can you please help me understand this further? Maybe I could work on this going ahead to simplify things here.

Reply:
Sure, the command-line on the receiving end (when GetActivatedEventArgs() parses it) already has a format that supports passing payload information.  Specifically I'm talking about the ability to pass a Uri that contains a key/value pair (the query string).

There are problems with the current implementation of GetActivatedEventArgs() in that it doesn't tolerate malformed command-lines well.  I have a change out to fix that, but it changes the expectations of the format as well.  

I'm mostly suggesting here that we should try to find a better way to utilize the command-line.  One that doesn't have the calling developer designing custom formats to support payloads.  We should instead find a way to utilize the Uri format and serialize + encode data into that format.  

This also lends itself to creating shared functions in the code base that generate that command-line.  I've started to do that in a super simple way with GenerateCommandLine() inside of ActivationRegistrationManager.cpp.  It could be moved and enhanced or accompanied with other helper functions to make a useful reusable toolset for callers to create command-lines in a consistent way.

I don't think any of this is the end of the world if not done.  At this point in the conversation though, it probably would be intentional technical debt.  Obviously we should avoid that when possible, but dealing with technical debt isn't the only factor in our lives. :)